res/res_musiconhold.c

```diff
@@ -1762,10 +1762,13 @@ static int local_ast_moh_start(struct ast_channel *chan, const char *mclass, con
 
 			if (ast_test_flag(global_flags, MOH_CACHERTCLASSES)) {
 				/* CACHERTCLASSES enabled, let's add this class to default tree */
+				ast_channel_lock(chan);
+				state = ast_channel_music_state(chan);
```

Comment:
Can we just do this (and the unlock) one time at the top of the function? Or do we need to serialize all access to `state`?

Reply:
I suspect there's potentially some long-running sub-tasks which is why I went a bit more granular.

Reply:
Long-running sub tasks like what?

Reply:
Good question.  You know better than me.  I didn't dig into this much further beyond this specific PR, life happened and I deviated from this.  My gut was that the construction of some of these generators could be time-consuming, and that that could have adverse impact on the channel structure.

But yea, if any of those do take overly long, it's going to be the case that the channel gets no audio anyway, and it would most likely be signalling stuff that just waits for a "long" duration, and that's probably OK if those gets blocked for a few hundred ms.

Based on that, sure, I think I can resubmit with "lock at top, release at bottom" and we see what happens?

Reply:
If memory serves, when this originally came up I did an audit of this file and I _think_ that the channel lock is serializing access to members of the `moh_files_state` structure, so the way it is here in this PR is probably the safest approach.

It would be helpful to go through and note which functions are being called with the channel lock already held, but that may be out of scope for this PR.

Reply:
> It would be helpful to go through and note which functions are being called with the channel lock already held, but that may be out of scope for this PR.

I went ahead and did this and pushed an additional commit that you can throw away or merge - up to you.

After your change, the only place in the module where the music state is manipulated without the channel lock held is in `local_ast_moh_cleanup` which is only called during channel destruction (but `ast_moh_cleanup()` is currently API so anyone _could_ call it).

I think locking around state manipulation in `local_ast_moh_cleanup` is technically the right thing, and the channel lock is still available at the time it is called, so it should be safe, and I've added another change to this PR that does that.